sinks/grpsink/grpsink.go

```diff
@@ -0,0 +1,164 @@
+func NewGRPCSpanSink(ctx context.Context, target, name string, commonTags map[string]string, log *logrus.Logger, opts ...grpc.DialOption) (*GRPCSpanSink, error) {
+	name = "grpc-" + name
+	conn, err := grpc.DialContext(ctx, target, opts...)
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			"name":   name,
+			"target": target,
+		}).Error("Error establishing connection to gRPC server")
+		return nil, err
+	}
+
+	return &GRPCSpanSink{
+		grpcConn:   conn,
+		name:       name,
+		target:     target,
+		commonTags: commonTags,
+		log:        log,
+	}, nil
+}
+
+// Start performs final preparations on the sink before it is
+// ready to begin ingesting spans.
+func (gs *GRPCSpanSink) Start(cl *trace.Client) error {
+	gs.traceClient = cl
+
+	// Run a background goroutine to do a little bit of connection state
+	// tracking.
+	go func() {
+		for {
+			// This call will block on a channel receive until the gRPC connection
+			// state changes. When it does, flip the marker over to allow another
+			// error to be logged from Ingest().
+			gs.grpcConn.WaitForStateChange(context.Background(), gs.grpcConn.GetState())
+			atomic.StoreUint32(&gs.loggedSinceTransition, 0)
+		}
+	}()
+	return nil
+}
+
+// Name returns this sink's name. As the gRPC sink is generic, it's expected
+// that this is set via configuration and injected.
+func (gs *GRPCSpanSink) Name() string {
+	return gs.name
+}
+
+// Ingest takes in a span and streams it over gRPC to the connected server.
+func (gs *GRPCSpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+
+	// Apply any common tags, superseding defaults passed in at sink creation
+	// in the event of overlap.
+	for k, v := range gs.commonTags {
+		ssfSpan.Tags[k] = v
+	}
+
+	client := NewSpanSinkClient(gs.grpcConn)
+	_, err := client.SendSpan(context.Background(), ssfSpan)
+
+	if err != nil {
+		atomic.AddUint32(&gs.dropCount, 1)
+
+		// gRPC guarantees that an error returned from an RPC call will be of
+		// type status.Status. In the unexpected event that they're not, this
+		// call creates a dummy type, so there's no risk of panic.
+		serr := status.Convert(err)
+		state := gs.grpcConn.GetState()
+
+		// Log all errors that occur in Ready state - that's weird. Otherwise,
+		// Log only one error per underlying connection state transition. This
+		// should be a reasonable heuristic to get an indicator that problems
+		// are occurring, without resulting in massive log spew while
+		// connections are under duress.
+		if state == connectivity.Ready || atomic.CompareAndSwapUint32(&gs.loggedSinceTransition, 0, 1) {
+			gs.log.WithFields(logrus.Fields{
+				logrus.ErrorKey: err,
+				"target":        gs.target,
+				"name":          gs.name,
+				"chanstate":     state.String(),
+				"code":          serr.Code(),
+				"details":       serr.Details(),
+				"message":       serr.Message(),
+			}).Warn("Error sending span to gRPC sink target")
```

Comment:
Should this be Warn or Error?

Reply:
hmm...yeah, errors are non-fatal, so this should probably be error. good call.

/me just noticed that it literally says `Warn("Error`